### PR TITLE
Fix for infinite iterator

### PR DESCRIPTION
**Pull request by dirkgr**
I changed this to use some code I've been using for a long time.

If we want a more minimal fix, I think this would do it as well:
```
def lazy_groups_of(iterable: Iterable[A], group_size: int) -> Iterable[List[A]]:
    iterable = iter(iterable)
    return iter(lambda: list(islice(iterable, 0, group_size)), [])
```

I find the long form more readable, but I'm happy either way.

### CODE REVIEW

File: allennlp/common/util.py [hunk 1]
```diff
@@ -106,12 +106,18 @@ def group_by_count(iterable: List[Any], count: int, default_value: Any) -> List[
 A = TypeVar("A")
 
 
-def lazy_groups_of(iterator: Iterator[A], group_size: int) -> Iterator[List[A]]:
+def lazy_groups_of(iterable: Iterable[A], group_size: int) -> Iterable[List[A]]:
     """
     Takes an iterator and batches the individual instances into lists of the
```
Comment (brendan-ai2) [inline allennlp/common/util.py]: s/iterator/iterable/

File: allennlp/common/util.py [hunk 2]
```diff
@@ -106,12 +106,18 @@ def group_by_count(iterable: List[Any], count: int, default_value: Any) -> List[
 A = TypeVar("A")
 
 
-def lazy_groups_of(iterator: Iterator[A], group_size: int) -> Iterator[List[A]]:
+def lazy_groups_of(iterable: Iterable[A], group_size: int) -> Iterable[List[A]]:
```
Comment (brendan-ai2) [inline allennlp/common/util.py]: The return type here shouldn't be `Iterable` given that we're aiming for laziness. I think you want to use `Generator`.

Comment (brendan-ai2) [reply allennlp/common/util.py]: (Or just stick with `Iterator`. The fact that it's a generator probably shouldn't matter.)

File: allennlp/common/util.py [hunk 3]
```diff
@@ -106,12 +106,18 @@ def group_by_count(iterable: List[Any], count: int, default_value: Any) -> List[
...
     """
     Takes an iterator and batches the individual instances into lists of the
     specified size. The last list may be smaller if there are instances left over.
     """
-    return iter(lambda: list(islice(iterator, 0, group_size)), [])
+    iterable = iter(iterable)
```
Comment (brendan-ai2) [inline allennlp/common/util.py]: Nit: `iterator = iter(iterable)` The `islice` call below looks like a bug when operating on an arbitrary iterable.

Comment (dirkgr) [reply allennlp/common/util.py]: I like to keep the number of local variables down for debugging, but I see your point. Will make the change.

File: allennlp/common/util.py (hunk 2 above)
Comment (dirkgr) [reply allennlp/common/util.py]: A `Generator` is an `Iterator`. So to make the minimum number of promises to the caller, I'll change it back to `Iterator`.

**Pull request merged**